gateway/ln-gateway/src/bin/cln_extension.rs

```diff
@@ -0,0 +1,330 @@
+                        // by passing the HTLC to the interceptor in the plugin state
+                        let payload: HtlcAccepted = serde_json::from_value(value)?;
+                        Ok(plugin.state().lock().await.intercept_htlc(payload).await)
+                    });
+                    handle.await?
+                },
+            )
+            .dynamic() // Allow reloading the plugin
+            .start(interceptor.clone())
+            .await?
+        {
+            let config = plugin.configuration();
+            let socket = PathBuf::from(config.lightning_dir).join(config.rpc_file);
+            let client = ClnRpc::new(socket).await.expect("connect to ln_socket");
+
+            let dir = match plugin.option("gateway-cfg") {
+                Some(options::Value::String(workdir)) => {
+                    // FIXME: cln_plugin doesn't yet support optional parameters
+                    if &workdir == "default-dont-use" {
+                        panic!("gateway-cfg option missing")
+                    } else {
+                        PathBuf::from(workdir)
+                    }
+                }
+                _ => unreachable!(),
+            };
+
+            Ok(Self {
+                client: Arc::new(Mutex::new(client)),
+                task_group: TaskGroup::new(),
+                interceptor,
+                dir,
+            })
+        } else {
+            // TODO: Accurately define LightningError when building the plugin fails
+            Err(ClnRpcError::RpcServerError(LightningError(Some(0))))
+        }
+    }
+}
+
+#[tonic::async_trait]
+impl GatewayLightning for ClnRpcService {
+    async fn get_pub_key(
+        &self,
+        _request: tonic::Request<GetPubKeyRequest>,
+    ) -> Result<tonic::Response<GetPubKeyResponse>, Status> {
+        self.client
+            .lock()
+            .await
+            .call(cln_rpc::Request::Getinfo(
+                model::requests::GetinfoRequest {},
+            ))
+            .await
+            .map(|response| {
+                let pub_key = match response {
+                    cln_rpc::Response::Getinfo(model::responses::GetinfoResponse {
+                        id, ..
+                    }) => id,
+                    _ => panic!("Unexpected response from cln_rpc"),
+                };
+                tonic::Response::new(GetPubKeyResponse {
+                    pub_key: pub_key.serialize().to_vec(),
+                })
+            })
+            .map_err(|e| {
+                error!("cln getinfo returned error: {:?}", e);
+                Status::internal(e.to_string())
+            })
+    }
+
+    async fn pay_invoice(
+        &self,
+        request: tonic::Request<PayInvoiceRequest>,
+    ) -> Result<tonic::Response<PayInvoiceResponse>, Status> {
+        let PayInvoiceRequest {
+            invoice,
+            max_fee_percent,
+            max_delay,
+        } = request.into_inner();
+
+        self.client
+            .lock()
```

Comment:
Can this pay pay 2 invoices concurrently? Seems to me that any attempt to pay an invoice while an invoice is being paid would block right here on `.lock()`.

Reply:
Right, we can improve this by making rpc take a stream of invoices to pay. Lock once, to get a client, then spawn a thread to pay all invoices on the stream...

Reply:
I think we'd also need to actually make the RPC call to CLN in a task. Which might take some work because `ClnRpc` isn't `Clone`.

I guess this is the current behavior today so not a blocker, just want to mention it now that I'm looking at this code for the first time in a while.

Reply:
Inspired #1248 as the `pay_invoice` rpc spec. Implementations like this one would work in batch mode by default from client perspective, even if they have internal limitations like you mention